scripts/create-darwin-volume.sh

```diff
@@ -303,8 +303,16 @@ generate_mount_daemon() {
 <!DOCTYPE plist PUBLIC "-//Apple Computer//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
 <plist version="1.0">
 <dict>
-  <key>RunAtLoad</key>
+  <key>LaunchOnlyOnce</key>
   <true/>
+  <key>KeepAlive</key>
+  <dict>
+    <key>PathState</key>
+    <dict>
+      <key>$NIX_ROOT/store</key>
+      <false/>
+    </dict>
+  </dict>
```

Comment:
This part concerns me. What does it mean to try and "keep alive" the mount command? It runs, and then exits. And since the job is now marked `LaunchOnlyOnce` launchd won't relaunch it. My best guess as to what this does is it says "don't run the mounter if the store is already mounted", but I'm not sure what the goal there is. If the problem is that trying to mount it when it's already mounted fails, we should update the mount command instead to be idempotent.

If we didn't mark this as `LaunchOnlyOnce` then I could see this as attempting to remount automatically if the nix store unmounts, though I suspect that would get rather annoying. I am a bit concerned that this new job configuration means I can't use `launchctl kickstart` to request the volume be remounted. I assume the use of `LaunchOnlyOnce` is meant to avoid having it immediately remount the volume if the user unmounts it (or to try and mount in a loop if the mount command fails).

Reply:
I am pretty confused about that, also. It seems like PathState should be more of a run condition than a KeepAlive condition, don't you think? Using LaunchOnlyOnce actually unloads the service when it is done, which is the hacky way that the nix-daemon configuration works with OtherJobEnabled. The docs are quite clear that OtherJobEnabled is not actually about enabled jobs, but about loaded ones. The idea is that the nix-daemon starts after the darwin-store is unloaded.

On the plus side, that should mean that you should be able to `launchctl bootstrap system /Library/LaunchDaemons/org.nixos.darwin-store.plist` to your hearts content to retry mounting the store volume.

Reply:
I missed the fact that `OtherJobEnabled` was set to `false`. I am curious if launchd guarantees that it loads all of the launchdaemons before evaluating any of the `OtherJobEnabled` conditions, because if not then it could plausibly load the daemon first, and decide it should be launched before loading the store mounter.

Also the lack of documentation on `LaunchOnlyOnce` about it actually unloading the job makes me nervous about this approach as well.

Reply:
I agree. I would really prefer just `PathState`. If anyone else out there can help test, it would help a lot.

Reply:
Another concern here will be that, regardless of the other ergonomics of the method, the volume doesn't just have to be mounted before the daemon starts--it also needs to be mounted before macOS tries to load or run anything that might be stored on it (files open in a restoring editor app, the shell for restoring terminal tabs/windows).

This isn't a problem for unencrypted systems, but there is a race-condition when FileVault is enabled. Any mechanism change will need to get tested against this case. I had a fairly straightforward test before, which was to enable FileVault, install Nix, open a file from the store volume in TextEdit, restart, and see if the document reopens.

Reply:
From what I recall, we never got that "it must mount before GUI" to actually be foolproof. The problem being that fstab wouldn't mount the volume before the GUI, and the LaunchDaemon that did this couldn't actually block the GUI, it would just cause the mount attempt to happen sooner and hopefully win. My recollection is that it mostly worked but sometimes failed, and that's why I actually still use an unencrypted Nix volume on my work laptop (my newer M1 laptop is using FileVault, but I've only restarted a handful of times and without a Nix-installed app running so there hasn't been much chance for the race to fail there).

Or is there something I'm missing about the current setup that actually makes it work?

Personally, it seems to me that we _should_ be able to use autofs to have it mount automatically when `/nix` is accessed. There's an old technical paper on autofs that indicates that `/etc/fstab` is actually used by autofs so I'm not sure why the mountpoints defined there _aren't_ mounted automatically upon access. If it did mount automatically upon access then we wouldn't expect to have an issue, because we could just attempt to launch `nix-daemon` and it would mount the volume for us. So the question is, why doesn't this work? And is there anything we haven't considered that would actually make it have the "mount automatically on access, blocking the operation until the mount completes" behavior? Because that would solve everything.

Reply:
> This isn't a problem for unencrypted systems, but there is a race-condition when FileVault is enabled. Any mechanism change will need to get tested against this case. We had a fairly straightforward test before, which was to enable FileVault, install Nix, open a file from the store volume in TextEdit, restart, and see if the document reopens.

Ok I tested this in a fresh vm: enabled filevault, opened TextEdit on a file in the store volume, restarted (with `reopen windows when logging back in` checked) and TextEdit was still there with the contents as before the reboot.

Reply:
> From what I recall, we never got that "it must mount before GUI" to actually be foolproof. The problem being that fstab wouldn't mount the volume before the GUI, and the LaunchDaemon that did this couldn't actually block the GUI, it would just cause the mount attempt to happen sooner and hopefully win. 

It's broadly accurate that we're depending on winning the race.

> My recollection is that it mostly worked but sometimes failed,

If anyone has receipts it'd be great to have them posted publicly, especially if it's reproducible. I had a reliable protocol for causing race-condition failures with the fstab-only mount, and when testing the daemon-mounted volume I was not once able to induce the same failure. I do not recall anyone telling me that they observed such a race-condition failure after using the test (and now release) installers. (But I'll leave room for eating crow since I'm also somewhat forgetful...) 

> Personally, it seems to me that we should be able to use autofs ... why doesn't this work? And is there anything we haven't considered that would actually make it have the "mount automatically on access, blocking the operation until the mount completes" behavior? Because that would solve everything.

I generally agree that this path sounded promising, but we flogged this pretty extensively back around https://github.com/NixOS/nix/pull/4181#issuecomment-717409562 through the end of that first PR. I've already expended days of scarce hobby time I can't get back banging my head against automount/autofs with nothing to show for it. As before, I'd be happy to have someone prove me wrong there--a simpler system solution would of course be better.

Reply:
I should have a vm with recovery mode available soon. I want to use that to look into automount. It definitely seems like the best way, but it seems like autofs has had some breaking changes the last few macos releases. I haven't been using darwin or nixos in that time, though. I'll keep you posted.